File: js/CjkDecoder.js
goog.provide('twim.CjkDecoder');

goog.require('twim.CjkTransform');

/**
 * @param{!Array<number>} data CKJ code-points
 * @return{!Uint8Array} bytes
 */
twim.CjkDecoder.decode = function(data) {
  let /** @type{number} */ nibblesLength = data.length * 2;
  let /** @type{!Uint16Array} */ nibbles = new Uint16Array(5 * Math.ceil(nibblesLength / 5) + 5);
  for (let /** @type{number} */ i = 0; i < data.length; ++i) {
    let /** @type{number} */ ord = twim.CjkTransform.unicodeToOrdinal(data[i] | 0);
    nibbles[2 * i] = ord % 296;
    nibbles[2 * i + 1] = ord / 296;
  }
  let /** @type{!Uint8Array} */ result = new Uint8Array(nibblesLength + Math.floor(nibblesLength / 38) + 6);
  let /** @type{number} */ resultLength = 0;
  let /** @type{number} */ ix = 0;
  let /** @type{number} */ acc = 0;
  let /** @type{number} */ bits = 0;
  while (ix < nibblesLength || acc !== 0) {
    if (bits < 8) {
      let /** @type{number} */ p = 0;
      for (let /** @type{number} */ j = 0; j < 5; ++j) {
        p = (p * 296) + (nibbles[ix + 4 - j] | 0);
      }
      ix += 5;
      acc += p * (1 << bits);
      bits += 41;
    }
    result[resultLength++] = acc & 0xFF;
    acc = Math.floor(acc / 256);
    bits -= 8;
  }
  while (result[resultLength - 1] === 0) {
    resultLength--;
  }
  return result.subarray(0, resultLength);
};

File: java/ru/eustas/twim/CjkDecoder.java
package ru.eustas.twim;

import java.io.ByteArrayOutputStream;

/**
 * CJK string to byte array decoder.
 */
final class CjkDecoder {

  // Collection of utilities. Should not be instantiated.
  CjkDecoder() {}

  static byte[] decode(int[] data) {
    int[] nibbles = new int[data.length * 2];
    for (int i = 0; i < data.length; ++i) {
      int ord = CjkTransform.unicodeToOrdinal(data[i]);
      nibbles[2 * i] = ord % 296;
      nibbles[2 * i + 1] = ord / 296;
    }
    ByteArrayOutputStream result = new ByteArrayOutputStream();
    int ix = 0;
    long acc = 0;
    int bits = 0;
    int zeroCount = 0;
    while (ix < nibbles.length || acc != 0) {
      if (bits < 8) {
        long p = 0;
        for (int j = 0; j < 5; ++j) {
          int idx = ix + 4 - j;
          int nibble = (idx < nibbles.length) ? nibbles[idx] : 0;
          p = (p * 296) + nibble;
        }
        ix += 5;
        acc += p << bits;
        bits += 41;
      }
      byte b = (byte)acc;
      if (b != 0) {
        while (zeroCount != 0) {
          result.write(0);
          zeroCount--;
        }
        result.write(b);
      } else {
        zeroCount++;
      }
      acc = acc >> 8;
      bits -= 8;
    }
    return result.toByteArray();
  }
}